contracts/smart-account/modules/AccountRecoveryModule.sol

```diff
@@ -0,0 +1,528 @@
+            recoveryThreshold,
+            securityDelay
+        );
+        for (uint256 i; i < length; ) {
+            if (guardians[i] == bytes32(0)) revert ZeroGuardian();
+            if (timeFrames[i].validUntil == 0)
+                timeFrames[i].validUntil = type(uint48).max;
+            if (timeFrames[i].validUntil < timeFrames[i].validAfter)
+                revert InvalidTimeFrame(
+                    timeFrames[i].validUntil,
+                    timeFrames[i].validAfter
+                );
+            if (timeFrames[i].validUntil < block.timestamp)
+                revert ExpiredValidUntil(timeFrames[i].validUntil);
+
+            _guardians[guardians[i]][msg.sender] = timeFrames[i];
+            emit GuardianAdded(msg.sender, guardians[i], timeFrames[i]);
+            unchecked {
+                ++i;
+            }
+        }
+        return address(this);
+    }
+
+    /**
+     * @dev validates userOps to submut and execute recovery requests
+     *     - if securityDelay is 0, it allows to execute the request immediately
+     *     - if securityDelay is non 0, the request is submitted and stored on-chain
+     *     - if userOp.callData matches the callData of the request already submitted,
+     *     - and the security delay has passed, it allows to execute the request
+     * @param userOp User Operation to be validated.
+     * @param userOpHash Hash of the User Operation to be validated.
+     * @return validation data (sig validation result + validUntil + validAfter)
+     */
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external virtual returns (uint256) {
+        // if there is already a request added for this userOp.callData, return validation success
+        // to procced with executing the request
+        // with validAfter set to the timestamp of the request + securityDelay
+        // so that the request execution userOp can be validated only after the delay
+        if (
+            keccak256(userOp.callData) ==
+            _smartAccountRequests[msg.sender].callDataHash
+        ) {
+            uint48 reqValidAfter = _smartAccountRequests[msg.sender]
+                .requestTimestamp +
+                _smartAccountSettings[msg.sender].securityDelay;
+            delete _smartAccountRequests[msg.sender];
+            return
+                VALIDATION_SUCCESS |
+                (0 << 160) | // validUntil = 0 is converted to max uint48 in EntryPoint
+                (uint256(reqValidAfter) << (160 + 48));
+        }
+
+        // otherwise we need to check all the signatures first
+        uint256 requiredSignatures = _smartAccountSettings[userOp.sender]
+            .recoveryThreshold;
+        if (requiredSignatures == 0) revert("AccRecovery: Threshold not set");
+
+        bytes calldata moduleSignature = userOp.signature[96:];
+
+        require(
+            moduleSignature.length >= requiredSignatures * 2 * 65,
```

Comment:
A comment about these magic numbers would be great

Reply:
+